### PR TITLE
[EMCAL-551] Move cell format to bitfield

### PR DESCRIPTION
**Pull request by mfasDa**
- Move underlying structure to bitfield and expand to
  48 bit (min. size)
- Increase number of time bits from 9 to 11 and make
  range signed in order to store negative times
- Add truncations for time and energy
- Add doxygen documentation for cell class

### CODE REVIEW

File: DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
Comment (shahor02) [inline DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h]: better avoid mixing root-style types like Short_t and internal uint16_t.

File: DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
Comment (shahor02) [inline DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h]: why do you use doubles? Even float precision is too much for your value.

File: DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
Comment (shahor02) [inline DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h]: 1st of all, root cannot store anonymous structs and you actually don't need such a union. 
Then, your struct has 2B alignment (largest data member) and the bit count of your data members are crossing 2B boundaries. So the compiler will add padding and as a result the sizeof(Cell) will be 8B instead of expected 6. With the bitset sized you have you need to use packig. Then, you need to init the vaulues.
So, you should use smt. like:

````
class Cell
{
 public:
  Cell() {
    auto tmp = reinterpret_cast<uint16_t*>(&mCellWord);
    tmp[0] = tmp[1] = tmp[2] = 0;
  }

  struct  __attribute__((packed)) CellWord {
      uint16_t mTowerID : 15;   ///< bits 0-14   Tower ID
      int16_t mTime : 11;          ///< bits 15-25: Time (signed, can become negative after calibration)
      uint16_t mEnergy : 14;    ///< bits 26-39: Energy
      uint16_t mCellStatus : 2; ///< bits 40-41: Cell1 status
      uint16_t mZerod : 6;      ///< bits 42-47: Zerod
    } mCellWord;

   ClassDefNV(Cell, 1);
};
````

File: DataFormats/Detectors/EMCAL/src/Cell.cxx
Comment (shahor02) [inline DataFormats/Detectors/EMCAL/src/Cell.cxx]: ``else if``

File: DataFormats/Detectors/EMCAL/src/Cell.cxx
Comment (shahor02) [inline DataFormats/Detectors/EMCAL/src/Cell.cxx]: again, why double?

File: DataFormats/Detectors/EMCAL/src/Cell.cxx
Comment (shahor02) [inline DataFormats/Detectors/EMCAL/src/Cell.cxx]: why this bitset is needed?

**Review by shahor02 — CHANGES_REQUESTED** [review]

File: DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
Comment (mfasDa) [reply DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h]: I see, OK, I will change this.

File: DataFormats/Detectors/EMCAL/src/Cell.cxx
Comment (mfasDa) [reply DataFormats/Detectors/EMCAL/src/Cell.cxx]: Sure, trivial.

File: DataFormats/Detectors/EMCAL/src/Cell.cxx
Comment (mfasDa) [reply DataFormats/Detectors/EMCAL/src/Cell.cxx]: Only for monitoring purpose, it is not really needed. I will remove the part.

File: DataFormats/Detectors/EMCAL/src/Cell.cxx
Comment (mfasDa) [reply DataFormats/Detectors/EMCAL/src/Cell.cxx]: In principle in order to avoid implicit type conversion (raw fitter will provide the time in double format). To be decided.

Comment (shahor02) [reply DataFormats/Detectors/EMCAL/src/Cell.cxx]: I don't think you need double at all, in the end, you fit everyting to <=15 bits.

**Comment by mfasDa** [comment]
@shahor02 I have tried to write and read cells from root files and I get 0 for all entries. Do I need to add something (i.e. in the linkdef file)?

**Comment by shahor02** [comment]
@mfasDa what data (raw or digits), workflow / options do you use?

**Comment by mfasDa** [comment]
I did the test with the run2->run3 converter here: https://github.com/mfasDa/EmcalRun3Converter/blob/master/EMCALRun3Converter/src/AliAnalysisTaskEmcalRun3ConverterCells.cxx

The task fills a vector with o2::emcal::Cell objects artificially combined to artificial timeframes and fills the vector to a tree. The writing step is what the CellWriterSpec would do as well.

I just inspected the tree for a fast check with a ROOT TBrowser and saw all entries 0, but also the file size is small for which I suspect the information is not streamed or only 0s are streamed.

**Comment by shahor02** [comment]
But this is what I wrote you about here: https://github.com/AliceO2Group/AliceO2/pull/4398 ? 
The decoding of raw data does not produce cells.

**Comment by mfasDa** [comment]
Can this work?

``` c++
class Cell {
...
 private:
  struct CellData {
    uint16_t mTowerID : 15;   ///< bits 0-14   Tower ID
    int16_t mTime : 11;       ///< bits 15-25: Time (signed, can become negative after calibration)
    uint16_t mEnergy : 14;    ///< bits 26-39: Energy
    uint16_t mCellStatus : 2; ///< bits 40-41: Cell status
    uint16_t mZerod : 6;      ///< bits 42-47: Zerod
  };

  CellData *getDataRepresentation() { return reinterpret_cast<CellData *>(mCellWords); }
  const CellData *getDataRepresentation() const { return reinterpret_cast<const CellData *>(mCellWords); }

  uint16_t mCellWords[3];     ///< data word

  ClassDefNV(Cell, 1);
};
```

I am just testing it.

**Comment by shahor02** [comment]
Hi,

No, to have the bits continuously packed you need to use ``__attribute__((packed))``

Cheers,
 Ruben

**Comment by mfasDa** [comment]
@shahor02 I tested the last implementation also with ROOT to file and I get reasonable distributions. The implementation seems to work.

**Comment by shahor02** [comment]
@mfasDa OK, will merge once CI is passed. Any news for raw to cells?

**Comment by mfasDa** [comment]
Please don't yet merge - we want to apply some change in the time representation in order to make better use of the time bits!

**Comment by mfasDa** [comment]
To clarify the situation on the time bits: With a readout window of 1.5 mus, with 11 bits we can cover 2 mus, so there is space for 500 ns unused. It would be possible to optimize the range by using a shift of 600 ns and make it all-positive (move from int16_t to uint16_t), in this case we could use a time resolution of 0.7 ns instead of 1 ns. 

Furthermore, as Martin pointed out, due to the requirement of 2-3 pre- and post-samples in the raw fit the effective time window is 1 mus. This would allow two more options:
a) reduce the amount of bits for time from 11 to 10 and stay with 1 ns time resolution. The free bit could then be assigned to improve the energy resolution.
b) Stay with 11 bits for time and go to 0.5 ns time resolution. 

Which of the 3 options we will take we will decide in the coming days. Up to then please do not merge the PR yet!

**Comment by mfasDa** [comment]
We decided on option1 with 11 bits and 1.5 mus readout window. @shahor02 the PR could now be merged.

**Review by sawenzel — CHANGES_REQUESTED** [review]
Looks like some EMCAL unit tests are failing. So either code or test needs fixes.

**Comment by mfasDa** [comment]
My bad, sorry - good that we have the test cases :) 

The reasons for the failing Cell test (all related to cell time) are now fixed. Note that due to precision change (from 1 ns to 0.73 ns) the check for the time stamp had to be changed from BOOST_CHECK_EQUAL to BOOST_CHECK_SMALL using the time resolution of the cell.

**Comment by mfasDa** [comment]
Finally build process are killed due to the analysis framework so people will hopefully start taking the enormous memory consumption of the analysis framework during the build process serious :)

**Comment by mfasDa** [comment]
@sawenzel @shahor02 Can you please have a look and eventually merge the PR? The failing test is not due to the commit but as commented above due to memory exceed in the analysis framework (an issue we are struggling with in building O2 on our machines since months). Super thanks!

**Review by shahor02 — APPROVED** [review]
Hi @mfasDa 

Looks ok now, merging. 
For the raw data -> cell decoding, did you check why the output is empty?

Cheers,
 Ruben

**Pull request merged**
